Task

bahubali2017 opened the pull request.

bahubali2017 committed:
Update README.md

bahubali2017 committed:
Update README.md

bahubali2017 committed:
Update README.md

bahubali2017 committed:
Please to readme file for the commmit details

bahubali2017 committed:
Update readme

bahubali2017 committed:
Added logger and root level="DEBUG"

adamklinkosz commented on src/main/java/pl/cleankod/exchange/entrypoint/model/ApiError.java:
```diff
@@ -1,4 +1,13 @@
+public record ApiError(int statusCode, String message, String details,LocalDateTime timestamp) {
+	// Constructor to initialize timestamp with the current time
+    public ApiError(int statusCode, String message, String details) {
+        this(statusCode, message, details, LocalDateTime.now()); 
```
Why do we need the current time for the API error?

bahubali2017 replied:
Initially I thought it will be useful to track the errors .But now I think it is now necessary and I am removing it.

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/AccountServiceImpl.java:
```diff
@@ -0,0 +1,60 @@
+import pl.cleankod.exchange.core.usecase.FindAccountUseCase;
+
+@Service
+public class AccountServiceImpl {
```
Why the `Impl` suffix?

bahubali2017 replied:
As this is part of the service Implementation layer.

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
         BigDecimal midRate = rateWrapper.rates().get(0).mid();
-        BigDecimal calculatedRate = money.amount().divide(midRate, RoundingMode.HALF_UP);
+        // we can choose RoundingMode.DOWN and scale =2  to avoid loosing money.
+        BigDecimal calculatedRate = money.amount().divide(midRate, 2, RoundingMode.DOWN);
```
This still suffers for potential money loss.

bahubali2017 replied:
Yes , I think we can do it the following way : Need your input
        // Perform division with higher precision and then round
        //RoundingMode.HALF_UP rounds up at halfway points, which can help retain value in calculations.
        BigDecimal calculatedRate = money.amount().divide(midRate, 4, RoundingMode.HALF_UP)
                                    .setScale(2, RoundingMode.HALF_UP);

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
+    public Money convert(Money money, Currency targetCurrency){
+    	 	
+        RateWrapper rateWrapper = getExchangeRate(targetCurrency);
+        LocalDateTime localDateTime = null;
```
Unused variable.

bahubali2017 replied:
Removed

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
+    @Cacheable(value = "currencyConversionCache", key = "#targetCurrency", unless = "#result == null")
+	public RateWrapper getExchangeRate(Currency targetCurrency) {
+    	
+		RateWrapper rateWrapper = exchangeRatesNbpClient.fetch("A", targetCurrency.getCurrencyCode());
```
Redundant variable.

bahubali2017 replied:
Removed

adamklinkosz commented on src/main/resources/ehCache.xml:
```diff
@@ -0,0 +1,18 @@
+        </expiry>
+        <resources>
+            <heap unit="entries">100</heap>
+            <offheap unit="MB">10</offheap>
```
What are the trade-offs of using off-heap memory, especially when we run this application inside a Docker container?

bahubali2017 replied:
Benefits of using off heap memory is reduced garbage collection as cached data is outside java heap .

But while  using docker ,we need to ensure the cache memory within container memory usage limit .Otherwise the application may crash by Docker runtime.

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
         this.exchangeRatesNbpClient = exchangeRatesNbpClient;
     }
 
-    @Override
```
Why remove this annotation?

bahubali2017 replied:
missed during file merge added it back

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
         return new Money(calculatedRate, targetCurrency);
     }
+    @Retry(name = "nbpApi")
+    @Cacheable(value = "currencyConversionCache", key = "#targetCurrency", unless = "#result == null")
```
Both mechanisms will not work. Can you spot why?

bahubali2017 replied:
These annotations use AOP to implement the functionality, so the annotated methods needed to be public  and  called from outside the object doing the retries or caching . Calling the method with the retry and caching logic from another method on the same class means the interceptor code doesn't execute, it has to be called from outside for the proxy to work.

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
     }
+    @Retry(name = "nbpApi")
+    @Cacheable(value = "currencyConversionCache", key = "#targetCurrency", unless = "#result == null")
+	public RateWrapper getExchangeRate(Currency targetCurrency) {
```
This breaks at least three of the S.O.L.I.D rules: `S`, `I` and potentially `L`. Can you spot why?

bahubali2017 replied:
1.Single Responsibility Principle -
The getExchangeRate method handles multiple responsibilities:

- Fetching the exchange rate from the NBP client.
- Handling retry logic in case of failure.
- Caching the result for subsequent requests.

Ideally, each of these should be separate concerns
2.ISP 
. If the RateWrapper contains more information than the convert method actually needs, then convert is forced to depend on something it doesn't fully use. If convert only needs the midRate, it shouldn't be forced to depend on the entire RateWrapper
3.LSP 
 If a subclass of RateWrapper exists (or is introduced later) with different behavior or constraints, it might cause unexpected issues within the convert method if it's substituted for the base RateWrapper. For example, a subclass might throw a different type of exception or return null under certain conditions that the base class doesn't.

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionStubService.java:
```diff
@@ -20,6 +20,6 @@ public Money convert(Money money, Currency targetCurrency) {
     private Money calculate(Money money, Currency targetCurrency) {
         BigDecimal rate = "PLN".equals(targetCurrency.getCurrencyCode()) ? EUR_TO_PLN_RATE : PLN_TO_EUR_RATE;
-        return Money.of(money.amount().multiply(rate).setScale(2, RoundingMode.HALF_UP), targetCurrency);
+        return Money.of(money.amount().multiply(rate).setScale(2, RoundingMode.HALF_DOWN), targetCurrency);
```
This rounding is different from the other one used in another part of the app. What do you suggest we do about it?

bahubali2017 replied:
We are planning to write like this Money.of(money.amount().multiply(rate).setScale(2, RoundingMode.HALF_UP), targetCurrency);

adamklinkosz commented on src/main/java/pl/cleankod/exchange/core/domain/Money.java:
```diff
@@ -1,26 +1,32 @@
+	
+	 @JsonCreator
+	  public Money(
+	            @JsonProperty("amount") BigDecimal amount,
```
The core domain model is now polluted by Jackson. Any ideas on how to remedy this?

bahubali2017 replied:
We can do any of the following -
1.implement custom serializers and deserializers that handle the JSON conversion without modifying the domain class itself.
2.using composition (wrapper class) to encapsulate serialization logic in separate classes.

adamklinkosz commented on src/main/java/pl/cleankod/exchange/core/domain/Account.java:
```diff
@@ -25,7 +41,7 @@ public static Id of(String value) {
         private static final Pattern PATTERN =
                 Pattern.compile("\\d{2}[ ]?\\d{4}[ ]?\\d{4}[ ]?\\d{4}[ ]?\\d{4}[ ]?\\d{4}[ ]?\\d{4}");
-
+        @JsonCreator
```
Core domain pollution happens here as well.

bahubali2017 replied:
yes agreed ....resolution is shared in above comment.

adamklinkosz commented on build.gradle:
```diff
@@ -34,3 +51,16 @@ dependencies {
+    dependsOn test
+    reports {
+        xml.required = true
+        csv.required = true
```
What do we need XML and CSV for?

bahubali2017 replied:
In  our current requirement we don't need it .Making both as false

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
-
 public class CurrencyConversionNbpService implements CurrencyConversionService {
+	
+	private final Logger logger = LoggerFactory.getLogger(CurrencyConversionNbpService.class);
```
Code formatting (here and in many other files you changed). I recommend using automatic code formatting hook that kicks in before a commit or turn that on in your IDE to avoid this kind of issues.

bahubali2017 replied:
Yes,agreed. I have implemented default formatting from Intellij IDE

adamklinkosz commented on src/main/java/pl/cleankod/exchange/provider/CurrencyConversionNbpService.java:
```diff
@@ -1,26 +1,50 @@
+        LocalDateTime localDateTime = null;
         BigDecimal midRate = rateWrapper.rates().get(0).mid();
-        BigDecimal calculatedRate = money.amount().divide(midRate, RoundingMode.HALF_UP);
+        // we can choose RoundingMode.DOWN and scale =2  to avoid loosing money.
```
I think this should be rather a commit message instead of a comment here. WDYT?

bahubali2017 replied:
Yes agreed

adamklinkosz reviewed (changes requested):
General issues:
* No test coverage for any of the introduced changes.
* Currently failing tests.
* Some provided solutions will simply not work (see code remarks for more details).
* Broken code quality principles.
* Many code formatting issues.
* Redundant code in some places.
* All changes introduced in one commit.

bahubali2017 committed:
1.Code formatting changes -Implemented IDE default formatting
2.In CurrencyConversionNbpService.java
  2.1  Perform division with higher precision and then round
        RoundingMode.HALF_UP rounds up at halfway points, which can help retain value in calculations.
  2,2 Removed unused , redundant variable added @override

bahubali2017 committed:
1.Added ExchangeRateService and  ExchangeRatesNbpClientWithRetry to
resolve violation of S,L, I in SOLID principle
.Also cache is implemented and working

2.ApiError  -removed un necessary LocalDateTime
3.updated ehcache key value pair

bahubali2017 committed:
1.imlplemented a solution(mapstruct)  to avoid value nested field in JSON
without  tinkering The core domain model
2.Removed unnecessary annotations and imports

bahubali2017 committed:
Removed annotation and formatting

bahubali2017 commented:
Update till now:-
1.No test coverage for any of the introduced changes.-WIP
2.Currently failing tests.-WIP
Groovy is new for me .I will try to write it .

3.Some provided solutions will simply not work (see code remarks for more details).-Done
 Retry mechanism -testing pending.

4.Broken code quality principles.-Done
5.Many code formatting issues.-Done
6.Redundant code in some places.-Done
All changes introduced in one commit.
